floors/enforce.go

```diff
@@ -0,0 +1,196 @@
+package floors
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+
+	"github.com/prebid/openrtb/v17/openrtb2"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/currency"
+	"github.com/prebid/prebid-server/exchange/entities"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+// EnforceFloors does floors enforcement for bids from all bidders based on floors provided in request, account level floors config
+func EnforceFloors(bidRequestWrapper *openrtb_ext.RequestWrapper, seatBids map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, account config.Account, conversions currency.Conversions) (map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, []error, []*entities.PbsOrtbSeatBid) {
+	rejectionsErrs := []error{}
+	rejecteBids := []*entities.PbsOrtbSeatBid{}
+
+	if isPriceFloorsDisabled(account, bidRequestWrapper) || isPriceFloorsEnforcementDisabledForRequest(bidRequestWrapper) {
+		return seatBids, []error{errors.New("Floors enforcement is disabled at account or in the request")}, rejecteBids
+	}
+
+	if !isFloorsSignallingSkipped(bidRequestWrapper) && isValidImpBidfloorPresentInRequest(bidRequestWrapper.BidRequest) {
+		if enforceFloors := isSatisfiedByEnforceRate(bidRequestWrapper, account.PriceFloors.EnforceFloorsRate, rand.Intn); enforceFloors {
+			enforceDealFloors := account.PriceFloors.EnforceDealFloors && getEnforceDealsFlag(bidRequestWrapper)
+
+			seatBids, rejectionsErrs, rejecteBids = enforceFloorToBids(bidRequestWrapper, seatBids, conversions, enforceDealFloors)
+		}
+	}
+	return seatBids, rejectionsErrs, rejecteBids
+}
+
+// enforceFloorToBids function does floors enforcement for each bid,
+// The bids returned by each partner below bid floor price are rejected and remaining eligible bids are considered for further processing
+func enforceFloorToBids(bidRequestWrapper *openrtb_ext.RequestWrapper, seatBids map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, conversions currency.Conversions, enforceDealFloors bool) (map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, []error, []*entities.PbsOrtbSeatBid) {
+	errs := []error{}
+	rejectedBids := []*entities.PbsOrtbSeatBid{}
+	impMap := make(map[string]*openrtb_ext.ImpWrapper, bidRequestWrapper.LenImp())
+
+	for _, v := range bidRequestWrapper.GetImp() {
+		impMap[v.ID] = v
+	}
+
+	for bidderName, seatBid := range seatBids {
+		eligibleBids := make([]*entities.PbsOrtbBid, 0, len(seatBid.Bids))
+		for _, bid := range seatBid.Bids {
+			retBid := checkDealsForEnforcement(bid, enforceDealFloors)
```

Comment:
Could this variable have a more descriptive name?

Reply:
renamed to `dealBid`